bundle-workflow/src/build_workflow/build_recorder.py

```diff
@@ -6,22 +6,84 @@
 
 import os
 import shutil
+from zipfile import ZipFile
 
 import yaml
+from jproperties import Properties  # type: ignore
 
 from manifests.build_manifest import BuildManifest
 
 
 class BuildRecorder:
+    class ArtifactInvalidError(Exception):
+        def __init__(self, path, message):
+            self.path = path
+            super().__init__(
+                f"Artifact {os.path.basename(path)} is invalid: {message}."
+            )
+
+    class PropertiesFile(Properties):
+        def __init__(self, filename, data):
+            super().__init__(self)
+            self.filename = filename
+            self.load(data)
+
+        def get_value(self, key, default_value=None):
+            try:
+                return self[key].data
+            except KeyError:
+                return default_value
+
+        def check_value(self, key, expected_value):
+            try:
+                value = self[key].data
+                if value != expected_value:
+                    raise BuildRecorder.ArtifactInvalidError(
+                        self.filename,
+                        f"expected to have {key}={expected_value}, but was {value}",
+                    )
+            except KeyError:
+                raise BuildRecorder.ArtifactInvalidError(
+                    self.filename,
+                    f"expected to have {key}={expected_value}, but none was found",
+                )
+
+        def check_value_in(self, key, expected_values):
+            try:
+                value = self[key].data
+                if value not in expected_values:
+                    raise BuildRecorder.ArtifactInvalidError(
+                        self.filename,
+                        f"expected to have {key}=any of {expected_values}, but was {value}",
+                    )
+            except KeyError:
+                if None not in expected_values:
+                    raise BuildRecorder.ArtifactInvalidError(
+                        self.filename,
+                        f"expected to have {key}=any of {expected_values}, but none was found",
+                    )
+
     def __init__(self, build_id, output_dir, name, version, arch, snapshot):
         self.output_dir = output_dir
+        self.version = str(version)
+        self.opensearch_version = (
+            self.version + "-SNAPSHOT" if snapshot else self.version
+        )
+        # BUG: the 4th digit is dictated by the component, it's not .0, this will break for 1.1.0.1
+        self.component_version = (
+            self.version + ".0-SNAPSHOT" if snapshot else f"{self.version}.0"
```

Comment:
can we replace SNAPSHOT with a variable QUALIFIER? It could be an "RC" candidate too in future

Reply:
That won't be enough to support build with a qualifier, I opened https://github.com/opensearch-project/opensearch-build/issues/385 for that.